smac/epm/partial_sparse_gaussian_process.py

```diff
@@ -0,0 +1,874 @@
+        """
+        if not self.optimize_kernel_hps:
+            raise ValueError("inducing_points will only be inactivate if self.optimize_kernel_hps is set True")
+        for p_name, t in self.named_parameters():
+            if p_name == 'covar_module.inducing_points':
+                t.requires_grad = False
+            else:
+                t.requires_grad = True
+
+    def forward(self, x):
+        if self.training:
+            if self.optimize_kernel_hps:
+                covar_x = self.base_covar(x)
+                mean_x = self._mean_module(x)
+            else:
+                covar_x = self.covar_module(x)
+                mean_x = self.mean_module(x)
+        else:
+            covar_x = self.covar_module(x)
+            mean_x = self.mean_module(x)
+        return MultivariateNormal(mean_x, covar_x)
+
+
+class VariationalGPModel(gpytorch.models.ApproximateGP):
+    """
+    A variational GP to compute the position of the inducing points
+    """
+    def __init__(self, kernel: Kernel, inducing_points: torch.tensor):
+        """
+        Initialize a Variational GP
+        Parameters:
+        ----------
+        kernel: Kernel
+            kernel of the variational GP, its hyperparameter needs to be fixed when it is used for initializing a PSGP
+        inducing_points: torch.tensor
+            inducing points
+        """
+        variational_distribution = gpytorch.variational.TrilNaturalVariationalDistribution(inducing_points.size(0))
+        variational_strategy = gpytorch.variational.VariationalStrategy(
+            self, inducing_points, variational_distribution, learn_inducing_locations=True
+        )
+        super(VariationalGPModel, self).__init__(variational_strategy)
+        self.mean_module = gpytorch.means.ZeroMean()
+        self.covar_module = kernel
+
+        shape_inducing_points = inducing_points.shape
+        lower_inducing_points = torch.zeros([shape_inducing_points[-1]]).repeat(shape_inducing_points[0])
+        upper_inducing_points = torch.ones([shape_inducing_points[-1]]).repeat(shape_inducing_points[0])
+
+        self.variational_strategy.register_constraint(param_name="inducing_points",
+                                                      constraint=Interval(lower_inducing_points,
+                                                                          upper_inducing_points,
+                                                                          transform=None),
+                                                      )
+        self.double()
+
+        for p_name, t in self.named_hyperparameters():
+            if p_name != "variational_strategy.inducing_points":
+                t.requires_grad = False
+
+    def forward(self, x):
+        mean_x = self.mean_module(x)
+        covar_x = self.covar_module(x)
+        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
+
+
+class PartialSparseGaussianProcess(GaussianProcessGPyTorch):
```

Comment:
Please make sure you also take the changes I suggested in the super class into account here.